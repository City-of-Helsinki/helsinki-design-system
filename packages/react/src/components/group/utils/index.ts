import {
  SyntheticEvent,
  FunctionComponent,
  HtmlHTMLAttributes,
  PropsWithChildren,
  ComponentType,
  MutableRefObject,
} from 'react';

import { StorageData, Storage } from './storage';

export type ChangeEventPayload = { value?: unknown; originalEvent?: SyntheticEvent };
export type ChangeEvent = { id: string; type?: string; payload?: ChangeEventPayload };
export type ChangeHandlerProps = ChangeEvent & { controller: Controller };
export type ChangeHandler = (props: ChangeHandlerProps) => unknown;
export type HTMLElementAttributesWithChildren = PropsWithChildren<HtmlHTMLAttributes<HTMLElement>>;
export type DefaultGroupElementProps = Record<string, unknown> & {
  key?: string;
  forwardedController?: Controller;
  'data-hds-group-id'?: string;
};

export type ElementRef = MutableRefObject<HTMLElement | null>;

export type Controller = {
  getProps: ({ id: string, elementProps: DefaultGroupElementProps }) => ReturnType<PropSetter>;
  updateData: (props: { data: StorageData; updateKeys?: string[] }) => void;
  getData: Storage['get'];
  getMetaData: () => StorageData;
  updateMetaData: (appended: StorageData) => StorageData;
  triggerChange: (change: ChangeEvent) => unknown;
};

export type ControllerProps = {
  reRenderer: () => void;
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange: ChangeHandler;
};

export type PropSetter<T = DefaultGroupElementProps> = (props: {
  id: string;
  controller: Controller;
  key: string;
  elementProps: DefaultGroupElementProps;
}) => T;

export type PropHandler = {
  getProps: (props: {
    id: string;
    controller: Controller;
    elementProps: DefaultGroupElementProps;
  }) => DefaultGroupElementProps;
  updateKey: (id: string) => void;
};

export type GroupChildFunction = FunctionComponent<{ controller: Controller }>;

export type GroupChild<P = DefaultGroupElementProps> = ComponentType<HTMLElementAttributesWithChildren & P>;

export type GroupProps = PropsWithChildren<unknown> & {
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange?: ChangeHandler;
};
