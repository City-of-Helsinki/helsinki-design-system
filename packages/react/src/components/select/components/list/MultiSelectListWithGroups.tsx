import React, { RefObject } from 'react';

import styles from '../../Select.module.scss';
import classNames from '../../../../utils/classNames';
import { SelectDataHandlers, SelectMetaData, DivElementProps, Group, SelectData } from '../../types';
import { getSelectedOptionsPerc, getGroupLabelOption, getVisibleGroupLabels, countVisibleOptions } from '../../utils';
import { MemoizedMultiSelectOption } from './listItems/MultiSelectOption';
import { useSelectDataHandlers } from '../../hooks/useSelectDataHandlers';
import { MultiSelectGroupLabel } from './listItems/MultiSelectGroupLabel';
import { getTextFromDataHandlers } from '../../texts';

const getGroupLabelIntermediateState = (target: Group): boolean => {
  const perc = getSelectedOptionsPerc(target);
  return perc < 1 && perc > 0;
};
const getGroupLabelDisabledState = (target: Group): boolean => {
  return !target.options.some((opt) => !opt.isGroupLabel && !opt.disabled);
};

export const createGroupOptionElements = (
  group: Group,
  { trigger, getOptionId }: Pick<SelectDataHandlers, 'trigger'> & Pick<SelectMetaData, 'getOptionId'>,
) => {
  return group.options
    .map((option) => {
      if (!option.visible) {
        return null;
      }

      if (option.isGroupLabel) {
        return (
          <MultiSelectGroupLabel
            option={option}
            trigger={trigger}
            getOptionId={getOptionId}
            isIntermediate={getGroupLabelIntermediateState(group)}
            isGroupDisabled={getGroupLabelDisabledState(group)}
            key={option.value}
          />
        );
      }
      return (
        <MemoizedMultiSelectOption
          option={option}
          trigger={trigger}
          isInGroup
          key={option.value}
          getOptionId={getOptionId}
        />
      );
    })
    .filter((option) => !!option);
};

export function MultiSelectGroup(props: DivElementProps) {
  const { children, ...attr } = props;
  return <div {...attr}>{children}</div>;
}

export const createGroupProps = (group: Group) => {
  const labelOption = getGroupLabelOption(group);
  return {
    role: 'group',
    'aria-label': (labelOption && labelOption.label) || '',
    key: labelOption && labelOption.label,
  };
};

export const createGroups = ({
  groups,
  getOptionId,
  trigger,
}: Pick<SelectData, 'groups'> & Pick<SelectMetaData, 'getOptionId'> & Pick<SelectDataHandlers, 'trigger'>) => {
  return groups.map((group) => {
    const attr = createGroupProps(group);
    const children = createGroupOptionElements(group, { trigger, getOptionId });
    if (!children.length) {
      return null;
    }
    return (
      <MultiSelectGroup {...attr} key={attr['aria-label']}>
        {children}
      </MultiSelectGroup>
    );
  });
};

export const createContainerProps = (
  dataHandlers: SelectDataHandlers,
): DivElementProps & { ref: RefObject<HTMLDivElement> } => {
  const { getData, getMetaData } = dataHandlers;
  const { groups } = getData();
  const { elementIds, refs, listInputType } = getMetaData();
  const hasInput = !!listInputType;
  const hasVisibleGroupLabels = getVisibleGroupLabels(groups).length > 0;
  const hasInputAndGroups = hasInput || hasVisibleGroupLabels;
  const hasOnlyGroups = !hasInput && hasVisibleGroupLabels;
  const choiceCount = countVisibleOptions(groups);
  const label = getTextFromDataHandlers('label', dataHandlers);
  const getRole = (): DivElementProps['role'] => {
    if (hasOnlyGroups) {
      return undefined;
    }
    if (hasInputAndGroups) {
      return 'dialog';
    }
    return 'listbox';
  };
  const getAriaLabel = (): DivElementProps['aria-labelledby'] => {
    if (hasOnlyGroups) {
      return undefined;
    }
    if (hasInputAndGroups) {
      return `${choiceCount} choices.`;
    }
    return `${label}. ${choiceCount} choices.`;
  };
  return {
    'aria-label': getAriaLabel(),
    id: elementIds.list,
    className: classNames(styles.list, styles.shiftOptions, styles.multiSelectList),
    ref: refs.list as unknown as RefObject<HTMLDivElement>,
    role: getRole(),
    tabIndex: -1,
  };
};

export function MultiSelectListWithGroups() {
  const dataHandlers = useSelectDataHandlers();
  const { getData, getMetaData, trigger } = dataHandlers;
  const { open, groups } = getData();
  const { isSearching, getOptionId } = getMetaData();
  const attr = createContainerProps(dataHandlers);

  const shouldRenderOptions = open && !isSearching;
  const children = shouldRenderOptions ? createGroups({ groups, getOptionId, trigger }) : [];
  return <div {...attr}>{open ? children : null}</div>;
}
