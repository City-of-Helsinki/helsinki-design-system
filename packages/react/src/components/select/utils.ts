import React, { ReactElement, ReactNode } from 'react';

import {
  SelectData,
  Group,
  SelectProps,
  Option,
  OptionInProps,
  SelectMetaData,
  GroupInProps,
  FilterFunction,
} from './types';
import { getChildrenAsArray } from '../../utils/getChildren';
import { ChangeEvent } from '../dataProvider/DataContext';
import { eventTypes } from './events';

type DomHandlerProps = {
  id: string;
  type?: string;
  trigger: (event: ChangeEvent) => void;
};

export type OptionIterator = (
  option: Option,
  group: Group,
  optionIndex: number,
  groupIndex: number,
) => Option | undefined;

export const DROPDOWN_MENU_ITEM_HEIGHT = 52;

export function createOnClickListener(props: DomHandlerProps) {
  const { id, type = eventTypes.click, trigger } = props;
  return {
    onClick: (originalEvent: React.MouseEvent) => {
      trigger({ id, type, payload: { originalEvent } });
    },
  };
}

export function createInputOnChangeListener(props: DomHandlerProps) {
  const { id, type = eventTypes.change, trigger } = props;
  return {
    onChange: (originalEvent: React.ChangeEvent<HTMLInputElement>) => {
      trigger({
        id,
        type,
        payload: { value: originalEvent.currentTarget.value, originalEvent },
      });
    },
  };
}

export function getOptionGroupIndex(groups: SelectData['groups'], option: OptionInProps): number {
  if (groups.length === 0) {
    return -1;
  }
  if (groups.length === 1) {
    return 0;
  }
  return groups.findIndex(({ options }) => {
    return (
      options.findIndex(({ value, isGroupLabel }) => value === option.value && isGroupLabel === option.isGroupLabel) >
      -1
    );
  });
}

export function iterateAndCopyGroup(groups: Group[], iterator: OptionIterator): Group[] {
  return groups.map((group, groupIndex) => {
    return {
      options: group.options.map((opt, optionIndex) => {
        return iterator(opt, group, optionIndex, groupIndex) || { ...opt };
      }),
    };
  });
}

export function getSelectedOptionsPerc(group: Group, pendingSelectionCount = 0): number {
  const optionCountWithoutGroupLabel = group.options.length - 1;
  if (!optionCountWithoutGroupLabel) {
    return 0;
  }
  return (
    (group.options.filter((option) => !option.isGroupLabel && option.selected).length + pendingSelectionCount) /
    optionCountWithoutGroupLabel
  );
}

export function updateOptionInGroup(
  groups: SelectData['groups'],
  updatedOption: Option,
  isMultiSelect: boolean,
): SelectData['groups'] {
  if (updatedOption.isGroupLabel) {
    throw new Error('Use updateGroupLabelAndOptions to update groupLabel and its related options');
  }
  const groupIndex = getOptionGroupIndex(groups, updatedOption);
  return groups.map((group, index) => {
    const selectedOptionPerc =
      index === groupIndex && isMultiSelect ? getSelectedOptionsPerc(group, updatedOption.selected ? 1 : -1) : 0;
    return {
      options: group.options.map((option) => {
        if (option.isGroupLabel) {
          if (index === groupIndex && isMultiSelect) {
            return {
              ...option,
              selected: selectedOptionPerc === 1,
            };
          }
          return { ...option };
        }
        if (index === groupIndex && option.value === updatedOption.value) {
          return {
            ...updatedOption,
            selected: !!updatedOption.selected,
          };
        }
        return {
          ...option,
          selected: !isMultiSelect ? false : option.selected,
        };
      }),
    };
  });
}

export function updateGroupLabelAndOptions(groups: SelectData['groups'], updatedOption: Option): SelectData['groups'] {
  if (!updatedOption.isGroupLabel) {
    throw new Error('Use updateOptionInGroup to update non-groupLabel options.');
  }
  if (!updatedOption.visible) {
    throw new Error('Cannot click an group label that is not visible (without a label)');
  }
  const targetGroupIndex = getOptionGroupIndex(groups, updatedOption);
  if (targetGroupIndex < 0) {
    return groups;
  }

  return iterateAndCopyGroup(groups, (option, group, optionIndex, groupIndex) => {
    if (groupIndex !== targetGroupIndex) {
      return { ...option };
    }
    return option.visible && !option.disabled
      ? {
          ...option,
          selected: updatedOption.selected,
        }
      : { ...option };
  });
}

export function clearAllSelectedOptions(groups: SelectData['groups']): SelectData['groups'] {
  return iterateAndCopyGroup(groups, (option) => {
    return {
      ...option,
      selected: false,
    };
  });
}

export function getAllOptions(groups: SelectData['groups'], filterOutGroupLabels = true): Option[] {
  const options: Option[] = [];
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (filterOutGroupLabels && option.isGroupLabel) {
        return;
      }
      options.push(option);
    });
  });
  return options;
}

export function getVisibleGroupLabels(groups: SelectData['groups']): Option[] {
  return groups.map((group) => group.options[0]).filter((option) => option && option.label && option.visible);
}

export function getSelectedOptions(groups: SelectData['groups']): Option[] {
  return getAllOptions(groups).filter((option) => !!option.selected);
}

export function validateOption(option: OptionInProps | string): Option {
  if (typeof option === 'string') {
    return { value: option, label: option, selected: false, isGroupLabel: false, visible: true, disabled: false };
  }

  const label = option.label || option.value || '';
  const value = option.value || label;

  return {
    label,
    value,
    selected: !!option.selected,
    isGroupLabel: false,
    visible: typeof option.visible === 'boolean' ? option.visible : true,
    disabled: typeof option.disabled === 'boolean' ? option.disabled : false,
  };
}

export function createGroupLabel(label: string) {
  return { ...validateOption(label), isGroupLabel: true, visible: !!label, disabled: false };
}

export function propsToGroups(props: Pick<SelectProps, 'groups' | 'options'>): SelectData['groups'] | undefined {
  if (!props.groups && !props.options) {
    return undefined;
  }
  if (props.groups) {
    return (props.groups as Group[]).map((group: Group) => {
      const hasLabelOptionAlready = !!group.options[0].isGroupLabel;
      const groupOptions = group.options.map(validateOption) as Option[];
      if (hasLabelOptionAlready) {
        return {
          options: [group.options[0], ...groupOptions.slice(1)],
        };
      }
      const labelOption: Option = createGroupLabel((group as GroupInProps).label);

      return {
        options: [labelOption, ...groupOptions],
      };
    });
  }

  return [
    {
      options: [createGroupLabel(''), ...(props.options || []).map(validateOption)],
    },
  ];
}

export function defaultFilter(option: Option, filterStr: string) {
  return option.label.toLowerCase().indexOf(filterStr.toLowerCase()) > -1;
}

export function filterOptions(groups: SelectData['groups'], filterStr: string, filterFunc: FilterFunction) {
  const newGroupsWithFilteredOptions = iterateAndCopyGroup(groups, (option) => {
    if (option.isGroupLabel) {
      return { ...option };
    }
    return { ...option, visible: !filterStr || filterFunc(option, filterStr) };
  });

  // check if group label should be visible....
  newGroupsWithFilteredOptions.forEach((group) => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    const groupLabel = getGroupLabelOption(group);
    if (!groupLabel) {
      return;
    }
    groupLabel.visible = !!groupLabel.label && group.options.findIndex((opt) => !opt.isGroupLabel && opt.visible) > -1;
  });

  return newGroupsWithFilteredOptions;
}

export function childrenToGroups(children: SelectProps['children']): SelectData['groups'] | undefined {
  if (!children || typeof children !== 'object') {
    return undefined;
  }
  const childArray = getChildrenAsArray(children) as ReactElement[];
  if (!childArray.length) {
    return [{ options: [] }];
  }
  const hasOptionGroups = childArray[0].type === 'optgroup';
  const optionElementToOption = (optionEl: ReactNode | ReactElement) => {
    const props = (
      optionEl && typeof optionEl === 'object' ? (optionEl as ReactElement).props : {}
    ) as ReactElement<HTMLOptionElement>['props'];
    const label = String(props.children);
    const value = props && String(props.value);
    const selected = !!(props && props.selected);
    const disabled = !!(props && props.disabled);
    return validateOption({ label, value, selected, disabled });
  };
  if (hasOptionGroups) {
    return childArray.map((child) => {
      const optionElements = child.props.children;
      const options = optionElements ? getChildrenAsArray(optionElements).map(optionElementToOption) : [];
      const label = createGroupLabel(String(child.props.label));
      options.unshift(label);
      return { options };
    });
  }
  return [{ options: [createGroupLabel(''), ...childArray.map(optionElementToOption)] }];
}

export function createSelectedOptionsList(currentSelections: Option[], groups: SelectData['groups']) {
  const selections = getSelectedOptions(groups);
  const selectionsAsValues = new Set(selections.map((opt) => opt.value));
  const stillSelected = currentSelections.filter((opt) => selectionsAsValues.has(opt.value));
  const selectedValues = new Set(stillSelected.map((opt) => opt.value));
  return [...stillSelected, ...selections.filter((opt) => !selectedValues.has(opt.value))];
}

export function getElementIds(containerId: string): SelectMetaData['elementIds'] {
  return {
    container: containerId,
    button: `${containerId}-main-button`,
    list: `${containerId}-list`,
    clearButton: `${containerId}-clear-button`,
    arrowButton: `${containerId}-arrow-button`,
    label: `${containerId}-label`,
    selectionsAndListsContainer: `${containerId}-sl-container`,
    tagList: `${containerId}-tag-list`,
    searchOrFilterInput: `${containerId}-input-element`,
    searchOrFilterInputLabel: `${containerId}-input-element-label`,
    clearAllButton: `${containerId}-clear-all-button`,
    showAllButton: `${containerId}-show-all-button`,
  };
}

export function countVisibleOptions(groups: SelectData['groups']): number {
  let count = 0;
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (!option.isGroupLabel && option.visible) {
        count += 1;
      }
    });
  });
  return count;
}

export function getGroupLabelOption(group: Group): Option | undefined {
  const firstOption = group.options[0];
  return firstOption && firstOption.isGroupLabel ? firstOption : undefined;
}
