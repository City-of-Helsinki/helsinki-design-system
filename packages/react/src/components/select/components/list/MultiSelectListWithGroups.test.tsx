/* eslint-disable jest/no-conditional-expect */
import { axe } from 'jest-axe';

import {
  initTests,
  mockedContainer,
  skipAxeRulesExpectedToFail,
  testUtilAfterAll,
  testUtilBeforeAll,
} from '../../testUtil';
import { getAllOptions } from '../../utils';
import { MultiSelectListWithGroups } from './MultiSelectListWithGroups';

jest.mock('../Container', () => {
  return {
    __esModule: true,
    Container: () => mockedContainer(),
  };
});

describe('<MultiSelectListWithGroups />', () => {
  beforeAll(() => {
    testUtilBeforeAll(MultiSelectListWithGroups);
  });
  afterAll(() => {
    testUtilAfterAll();
  });
  const selectProps = { open: true, multiSelect: true };
  const initProps = {
    renderComponentOnly: true,
    selectProps,
    testProps: { groups: true },
  };
  it('Component is rendered', async () => {
    const { asFragment } = initTests(initProps);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = initTests(initProps);
    const results = await axe(container, skipAxeRulesExpectedToFail);
    expect(results).toHaveNoViolations();
  });
  it('all options are rendered and have correct attributes', async () => {
    const { getGroups, getOptionElement } = initTests(initProps);
    const allOptions = getAllOptions(getGroups(), false);
    const uniqueIds = new Set<string>();
    // eslint-disable-next-line no-restricted-syntax
    for (const option of allOptions) {
      const el = getOptionElement(option);
      expect(el.getAttribute('role')).toBe('checkbox');
      const id = el.getAttribute('id') as string;
      expect(!!id).toBeTruthy();
      expect(uniqueIds.has(id)).toBeFalsy();
      expect(el.getAttribute('tabindex')).toBe('-1');
      uniqueIds.add(id);
      expect(el.getAttribute('aria-disabled')).toBe('false');
      expect(el.getAttribute('aria-checked')).toBe('false');
    }
  });
  it('Clicking a non-group label option triggers an event and selects it. Group label changes too.', async () => {
    const {
      getGroups,
      clickOptionElement,
      getOnChangeCallArgsAsProps,
      getOnChangeMockCalls,
      createRenderUpdatePromise,
      getOptionElement,
    } = initTests({
      ...initProps,
      renderComponentOnly: false,
    });
    const groups = getGroups();
    let selectCount = 0;
    let clickedGroupOptionsCount = 0;
    // eslint-disable-next-line no-restricted-syntax
    for (const group of groups) {
      const { options } = group;
      clickedGroupOptionsCount = 0;
      const groupLabel = options[0];
      // eslint-disable-next-line no-restricted-syntax
      for (const option of options) {
        if (!option.isGroupLabel) {
          const renderUpdatePromise = createRenderUpdatePromise();
          const callCount = getOnChangeMockCalls().length;
          clickOptionElement(option);
          selectCount += 1;
          clickedGroupOptionsCount += 1;
          expect(getOnChangeMockCalls().length).toBe(callCount + 1);
          const onChangeProps = getOnChangeCallArgsAsProps();
          expect(onChangeProps.clickedOption).toMatchObject(option);
          // the option sent to the onChange has .selected === false
          expect(onChangeProps.clickedOption.selected).toBeFalsy();
          expect(onChangeProps.selectedOptions).toHaveLength(selectCount);
          // the option in the array of selected items has .selected === true
          expect(onChangeProps.selectedOptions[0].selected).toBeTruthy();
          // eslint-disable-next-line no-await-in-loop
          await renderUpdatePromise;
        }
        if (clickedGroupOptionsCount > 0) {
          const el = getOptionElement(groupLabel);
          const allSelected = clickedGroupOptionsCount === options.length - 1;
          expect(el.getAttribute('aria-checked')).toBe(allSelected ? 'true' : 'mixed');
          expect((el.querySelector('input') as HTMLInputElement).indeterminate).toBe(!allSelected);
        }
      }
    }
  });
  it('Clicking a group label option triggers an event and selects all options in that group.', async () => {
    const {
      getGroups,
      clickOptionElement,
      getOnChangeCallArgsAsProps,
      getOnChangeMockCalls,
      createRenderUpdatePromise,
    } = initTests({
      ...initProps,
      renderComponentOnly: false,
    });
    const groups = getGroups();
    let selectCount = 0;
    // eslint-disable-next-line no-restricted-syntax
    for (const group of groups) {
      const callCount = getOnChangeMockCalls().length;
      const { options } = group;
      const groupLabel = options[0];
      const renderUpdatePromise = createRenderUpdatePromise();
      clickOptionElement(groupLabel);
      // -1 because group labels are not in the list of selected options
      selectCount += options.length - 1;
      expect(getOnChangeMockCalls().length).toBe(callCount + 1);
      const onChangeProps = getOnChangeCallArgsAsProps();
      expect(onChangeProps.clickedOption).toMatchObject(groupLabel);
      // the option sent to the onChange has .selected === false
      expect(onChangeProps.clickedOption.selected).toBeFalsy();
      expect(onChangeProps.selectedOptions).toHaveLength(selectCount);
      // the option in the array of selected items has .selected === true
      expect(onChangeProps.selectedOptions[0].selected).toBeTruthy();
      // eslint-disable-next-line no-await-in-loop
      await renderUpdatePromise;
    }
  });
});
