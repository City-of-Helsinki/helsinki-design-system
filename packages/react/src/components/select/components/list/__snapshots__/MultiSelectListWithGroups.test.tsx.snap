// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSelectListWithGroups /> Component is rendered 1`] = `
<DocumentFragment>
  <div
    class="list shiftOptions multiSelectList"
    id="test-select-component-list"
    tabindex="-1"
  >
    <div
      aria-label="Group0"
      role="group"
    >
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="FI: Group0 (choose all)"
        class="listItem groupLabel focusableListItem selectableListItem multiSelectListItem"
        id="test-select-component-option-0"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-0-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-0-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-0-checkbox"
          >
            Group0
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group0_Option0_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-1"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-1-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-1-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-1-checkbox"
          >
            Group0_Option0_Label
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group0_Option1_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-2"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-2-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-2-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-2-checkbox"
          >
            Group0_Option1_Label
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group0_Option2_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-3"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-3-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-3-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-3-checkbox"
          >
            Group0_Option2_Label
          </label>
        </div>
      </div>
    </div>
    <div
      aria-label="Group1"
      role="group"
    >
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="FI: Group1 (choose all)"
        class="listItem groupLabel focusableListItem selectableListItem multiSelectListItem"
        id="test-select-component-option-4"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-4-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-4-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-4-checkbox"
          >
            Group1
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group1_Option0_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-5"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-5-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-5-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-5-checkbox"
          >
            Group1_Option0_Label
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group1_Option1_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-6"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-6-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-6-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-6-checkbox"
          >
            Group1_Option1_Label
          </label>
        </div>
      </div>
    </div>
    <div
      aria-label="Group2"
      role="group"
    >
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="FI: Group2 (choose all)"
        class="listItem groupLabel focusableListItem selectableListItem multiSelectListItem"
        id="test-select-component-option-7"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-7-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-7-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-7-checkbox"
          >
            Group2
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group2_Option0_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-8"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-8-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-8-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-8-checkbox"
          >
            Group2_Option0_Label
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group2_Option1_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-9"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-9-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-9-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-9-checkbox"
          >
            Group2_Option1_Label
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group2_Option2_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-10"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-10-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-10-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-10-checkbox"
          >
            Group2_Option2_Label
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group2_Option3_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-11"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-11-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-11-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-11-checkbox"
          >
            Group2_Option3_Label
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group2_Option4_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-12"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-12-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-12-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-12-checkbox"
          >
            Group2_Option4_Label
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group2_Option5_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-13"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-13-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-13-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-13-checkbox"
          >
            Group2_Option5_Label
          </label>
        </div>
      </div>
      <div
        aria-checked="false"
        aria-disabled="false"
        aria-label="Group2_Option6_Label"
        class="listItem selectableListItem focusableListItem multiSelectListItem"
        id="test-select-component-option-14"
        role="checkbox"
        tabindex="-1"
      >
        <div
          aria-hidden="true"
          class="checkboxContainer"
        >
          <div
            class="checkbox"
          >
            <input
              class="input"
              id="test-select-component-option-14-checkbox"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <label
              class="label noLabel"
              for="test-select-component-option-14-checkbox"
            />
          </div>
          <label
            for="test-select-component-option-14-checkbox"
          >
            Group2_Option6_Label
          </label>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
